src/main/java/io/supertokens/inmemorydb/ConnectionWithLocks.java

```diff
@@ -17,26 +17,28 @@
 package io.supertokens.inmemorydb;
 
 import java.sql.*;
-import java.util.ArrayList;
-import java.util.List;
+import java.util.HashSet;
 import java.util.Map;
 import java.util.Properties;
+import java.util.Set;
 import java.util.concurrent.Executor;
 
 public class ConnectionWithLocks implements Connection {
 
     private Connection con;
     private ConnectionPool connectionPool;
-    private List<String> lockedKeys = new ArrayList<>();
+    private Set<String> lockedKeys = new HashSet<String>();
 
     public void lock(String key) {
-        connectionPool.lock(key);
-        this.lockedKeys.add(key);
+        if (!this.lockedKeys.contains(key)) {
+            this.lockedKeys.add(key);
+            connectionPool.lock(key);
```

Comment:
This is guaranteed to work only of the same thread is going to be calling `lock` on the same instance of `ConnectionWithLocks ` (which I think is true).

If one `ConnectionWithLocks` instance is being shared by multiple threads (which I think never happens), there can still be a deadlock here cause two threads can try and lock the same `key` and get `this.lockedKeys.contains(key)` as `false`, thereby entering the `if` statement, and then both running `connectionPool.lock(key)`, which will hang that thread. In this case, we can solve this by making the `lock` function into `public synchronized void lock(String key) {` and same for `unlockAllLocks` function (it might be worth adding the `synchronized` keyword anyway.

Reply:
Hmm, I didn't look too deep at the rest of the class to see if it'd work while shared between threads. I think this wouldn't be misleading (saying this should work when shared) and it doesn't have too much overhead as far as I can see, so I'll add it.

Reply:
Alright!